### ISSUE TITLE
Uploading logs to GCS may break the whole run on Jenkins

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#13548]
In our scalability tests it sometimes happens that all tests are passing, but the run is marked red as "failure". This is because uploading logs to GCS sometimes failed with error:

    ---
    [Google Cloud Storage Uploader] Uploading: _artifacts/146.148.94.88:22-kubelet.log
    ERROR: [Google Cloud Storage Uploader] Exception while uploading files: exception uploading file

I've seen 2 examples during last 24 hours:
http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce-scalability/2319/
http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce-scalability/2325/

For now it's just misleading.

However, since scalability suite is now pretty stable, I would like to add it to list of projects that our merge bot requires to be green to merge anything. Unfortunately this looks like a blocker for it.

 Any thoughts how to fix that?

cc @kubernetes/goog-testing @quinton-hoole @ixdy

**Comment by wojtek-t** [comment]
cc @fgrzadkowski

**Comment by wojtek-t** [comment]
Another example of it:
http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce-scalability/2328/

**Comment by quinton-hoole** [comment]
Yes, I think this is a good idea.  It should be relatively trivial to mask the uploading error from Jenkins (in the script).  We do ideally need some sort of warning to alert us if/when the uploads are failing repeatedly (which could be done in a separate PR).  Ideally it would email the oncall, or automatically (re-)open and assign a github ticket.

@ixdy do you have time to look at this next week?  Anyone else?  If not, I will.  Tentatively assigning to myself, although hoping that someone else will pick it up.

**Comment by ixdy** [comment]
We're using a Jenkins plugin, not the e2e scripts, to upload build results to GCS. There's no way that I can find to make it not fail the build if an upload fails.

Side note: the actual error message is a 401 error from GCS
```
Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 401 Unauthorized
	at com.google.api.client.googleapis.json.GoogleJsonResponseException.from(GoogleJsonResponseException.java:145)
```
which seems to suggest an actual issue in the underlying infrastructure somewhere, which probably shouldn't be overlooked, but I have no idea where to start to try to figure which team to notify.

**Comment by ixdy** [comment]
That said, it should be pretty straightforward to write a script to upload the artifacts and build log, rather than relying on the plugin.

**Comment by wojtek-t** [comment]
Yes - that would be great if you could do it...

**Comment by wojtek-t** [comment]
Just bumping it up - it happened again few times over the weekend and also today:
http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce-scalability/2373/

**Comment by quinton-hoole** [comment]
I looked at the existing plugin that we're using, and it does indeed retry (5 times!)

https://github.com/jenkinsci/google-storage-plugin/blob/3fcd85a12cc6a6994355e50fedffe3015feecb26/src/main/java/com/google/jenkins/plugins/storage/AbstractUpload.java#L467

https://github.com/jenkinsci/google-storage-plugin/blob/913f85f9753f8b4869dff2ffb3c06891157cd593/src/main/java/com/google/jenkins/plugins/storage/UploadModule.java#L67


Also @ixdy searched the logs and found a lot of GCS errors:
$ grep GoogleJsonResponseException: jenkins.log | cut -d' ' -f2- | sort | uniq -c
    50 401 Unauthorized
     22 500 Internal Server Error
     12 503 Service Unavailable
So I don't really think that there is much more that we can do here other than help GCS to track down their errors, and increase the retry count (although 5 with exponential backoff already seems  rather excessive).

**Comment by wojtek-t** [comment]
Thanks for looking into it @quinton-hoole and @ixdy !

Is it easy to increase the retry count? Can we do it (I agree that 5 doesn't seem low, but we still observe errors).

Did you talk to GCS people? I wouldn't expect the number of errors to be such big...

**Comment by quinton-hoole** [comment]
No, the retry count is unfortunately not configurable.  I have not yet spoke  to the GCS people, but plan to.

**Comment by quinton-hoole** [comment]
@wojtek-t ^

**Comment by ixdy** [comment]
@quinton-hoole did we ever follow up with the GCS folks?

I'm working on a PR to upload test artifacts/logs all within e2e.sh itself.

**Comment by quinton-hoole** [comment]
@ixdy, sorry, no.

**Comment by wojtek-t** [comment]
@quinton-hoole - why did you close this? IIUC the PR is still not merged.

**Comment by quinton-hoole** [comment]
@wojtek-t Yes, sorry, you're right. I'll assign this one to @ixdy, to close once the PR is merged.

**Comment by ixdy** [comment]
I enabled the new script for -scalability, but it will take a few hours to verify that it works as intended.
I haven't disabled the plugin yet, but assuming everything works as intended, I'll do that during the day PDT Friday.

**Comment by wojtek-t** [comment]
@ixdy - thanks!

**Comment by wojtek-t** [comment]
@ixdy - this doesn't work at all - in fact the cluster doesn't even start after this change

**Comment by wojtek-t** [comment]
@ixdy - I reverted your change
Can you please experiment on something different than -scalability - if this doesn't work it blocks merge bot for ~2.5 hours which is how long a single run lasts.
Can you please experiment on something that is not blocking merge-bot first?

FYI: @piosz @quinton-hoole

**Comment by ixdy** [comment]
Ugh, sorry about that. It'd been working fine on a number of builds, but all of those cleaned their workspaces before running tests.

#15786 should fix the issue, and I'll test it on one of the flaky builds first.
